lib/hanami/action/flash.rb

```diff
@@ -107,112 +129,111 @@ def clear
       # @since 0.3.0
       # @api private
       def empty?
-        _values.all?(&:empty?)
+        _values.empty?
       end
 
       # @return [String]
       #
       # @since 1.0.0
       def inspect
-        "#<#{self.class}:#{'0x%x' % (__id__ << 1)} #{@merged.inspect}>"
+        "#<#{self.class}:#{'0x%x' % (__id__ << 1)} {:data=>#{flash.inspect}, :kept=>#{kept_data.inspect}} >"
       end
 
-      private
-
-      # The flash registry that holds the data for **all** the recent requests
-      #
-      # @return [Hash] the flash
+      # Set @keep to true, is use when storing flash and triggering a redirect, to store
+      # previous request flash.
+      # @return [TrueClass]
       #
-      # @since 0.3.0
+      # @since x.x.x
       # @api private
-      def flash
-        @session[SESSION_KEY] || {}
+      def keep!
+        return if empty?
+        @keep = true
       end
 
-      # The flash registry that holds the data **only for** the current request
+      private
+
+      # Get the kept request data
       #
-      # @return [Hash] the flash for the current request
+      # @return [Hash]
       #
-      # @since 0.3.0
+      # @since x.x.x
       # @api private
-      def data
-        flash[@request_id] || {}
+      def kept
+        @session[KEPT_KEY] || []
       end
 
-      # Expire the stale data from the previous request.
+      # Merge current data into KEPT_KEY hash
       #
-      # @return [void]
+      # @return [Hash] the current value of KEPT_KEY
       #
-      # @since 0.3.0
+      # @since x.x.x
       # @api private
-      def expire_stale!
-        flash.each do |request_id, _|
-          flash.delete(request_id) if delete?(request_id)
-        end
+      def keep_data
+        new_kept = kept << Kept.new(flash)
+
+        @session[KEPT_KEY] = new_kept
       end
 
-      # Remove the flash entirely from the session if empty.
+      # Return value for key inside KEPT_KEY hash and delete it.
       #
-      # @return [void]
+      # @return [String|Integer|Nil] the current value for that key
       #
-      # @since 0.3.0
+      # @since x.x.x
       # @api private
-      #
-      # @see Hanami::Action::Flash#empty?
-      def remove!
-        @session.delete(SESSION_KEY) if empty?
+      def expire_kept!
+        kept.reject!(&:expired?)
       end
 
-      # @since 1.0.0
-      # @api private
-      #
-      # @see Hanami::Action::Flash#[]
-      def merge!
-        @merged = last_request_flash.merge(data)
+
+      def update_kept_request_count
+        kept.each(&:increase_request_count!)
       end
 
-      # Values from all the stored requests
+      def search_in_kept_data(key)
+        value_in_kept = kept.find { |kept| kept.key?(key) }
+        return unless value_in_kept
+
+        value_in_kept.find(key)
```

Comment:
Can we change to:
```ruby
value_in_kept = kept.find { |kept| kept.key?(key) }
value_in_kept&.find(key)
```
Could even do one line, but I think that'd be confusing with double `find`.